File: packages/sdk-ts/src/telemetry.ts
/** Provider usage captured by wrapClient (Anthropic / OpenAI). */
export interface ProviderUsage {
  model: string;
  tokens_in: number;
  tokens_out: number;
  latency_ms: number;
  captured_at_ms: number;
}

export interface TelemetryConfig {
  /**
   * When true, emitEdge fills missing tokens/cost from previews + model pricing.
   * Env: BLAMR_ENRICH_USAGE=1 (default on).
   */
  enrichMissingUsage?: boolean;
  /**
   * When true, emitEdge consumes usage from the last wrapped LLM call if tokens are 0.
   * Env: BLAMR_ATTACH_PROVIDER_USAGE=1 (default on).
   */
  attachProviderUsage?: boolean;
  /** Rough chars-per-token for preview estimation (default 4). */
  charsPerToken?: number;
  /** Override or extend default model pricing (USD per 1M tokens). */
  modelPricing?: Record<string, { inputPer1M: number; outputPer1M: number }>;
}

const DEFAULT_CHARS_PER_TOKEN = 4;

/** USD per 1M tokens — extend via telemetry.modelPricing. */
const DEFAULT_MODEL_PRICING: Record<string, { inputPer1M: number; outputPer1M: number }> = {
  'claude-sonnet-4-6': { inputPer1M: 3, outputPer1M: 15 },
  'claude-3-5-sonnet-latest': { inputPer1M: 3, outputPer1M: 15 },
  'claude-3-5-sonnet-20241022': { inputPer1M: 3, outputPer1M: 15 },
  'claude-3-5-haiku-latest': { inputPer1M: 0.8, outputPer1M: 4 },
  'claude-3-opus-latest': { inputPer1M: 15, outputPer1M: 75 },
  'gpt-4o': { inputPer1M: 2.5, outputPer1M: 10 },
  'gpt-4o-mini': { inputPer1M: 0.15, outputPer1M: 0.6 },
  'gpt-4.1': { inputPer1M: 2, outputPer1M: 8 },
  'gpt-4.1-mini': { inputPer1M: 0.4, outputPer1M: 1.6 },
  'llama3.2:3b': { inputPer1M: 0, outputPer1M: 0 },
};

const CALL_TYPE_ALIASES: Record<string, string> = {
  tool_call: 'Tool call',
  llm_call: 'LLM call',
  mcp_call: 'MCP call',
  vision_call: 'Vision call',
};

export function resolveTelemetryConfig(cfg?: TelemetryConfig): Required<TelemetryConfig> {
  const envEnrich =
    typeof process !== 'undefined' && process.env?.BLAMR_ENRICH_USAGE !== undefined
      ? process.env.BLAMR_ENRICH_USAGE !== '0' && process.env.BLAMR_ENRICH_USAGE !== 'false'
      : cfg?.enrichMissingUsage ?? true;
  const envAttach =
    typeof process !== 'undefined' && process.env?.BLAMR_ATTACH_PROVIDER_USAGE !== undefined
      ? process.env.BLAMR_ATTACH_PROVIDER_USAGE !== '0' && process.env.BLAMR_ATTACH_PROVIDER_USAGE !== 'false'
      : cfg?.attachProviderUsage ?? true;
  return {
    enrichMissingUsage: cfg?.enrichMissingUsage ?? envEnrich,
    attachProviderUsage: cfg?.attachProviderUsage ?? envAttach,
    charsPerToken: cfg?.charsPerToken ?? DEFAULT_CHARS_PER_TOKEN,
    modelPricing: { ...DEFAULT_MODEL_PRICING, ...cfg?.modelPricing },
  };
}

function estimateTokens(text: string | undefined, charsPerToken: number): number {
  if (!text?.trim()) return 0;
  return Math.max(1, Math.ceil(text.trim().length / charsPerToken));
}

function resolvePricing(
  model: string,
  pricing: Record<string, { inputPer1M: number; outputPer1M: number }>,
): { inputPer1M: number; outputPer1M: number } | null {
  if (!model || model === 'unknown') return null;
  if (pricing[model]) return pricing[model];
  const lower = model.toLowerCase();
  for (const [key, value] of Object.entries(pricing)) {
    if (lower.includes(key.toLowerCase()) || key.toLowerCase().includes(lower)) {
      return value;
    }
  }
  if (lower.includes('sonnet')) return { inputPer1M: 3, outputPer1M: 15 };
  if (lower.includes('haiku')) return { inputPer1M: 0.8, outputPer1M: 4 };
  if (lower.includes('opus')) return { inputPer1M: 15, outputPer1M: 75 };
  if (lower.includes('gpt-4o-mini')) return { inputPer1M: 0.15, outputPer1M: 0.6 };
  if (lower.includes('gpt-4')) return { inputPer1M: 2.5, outputPer1M: 10 };
  return null;
}

export function estimateCostUsd(
  model: string,
  tokensIn: number,
  tokensOut: number,
  pricing: Record<string, { inputPer1M: number; outputPer1M: number }>,
): number {
  const rates = resolvePricing(model, pricing);
  if (!rates) return 0;
  const cost = (tokensIn * rates.inputPer1M + tokensOut * rates.outputPer1M) / 1_000_000;
  return Math.round(cost * 1_000_000) / 1_000_000;
}

export function normalizeCallType(callType: string | undefined, model: string): string {
  const raw = (callType ?? 'LLM call').trim();
  const lower = raw.toLowerCase();
  if ((lower === 'tool_call' || lower === 'llm_call') && model && model !== 'unknown') {
    return 'LLM call';
  }
  const aliased = CALL_TYPE_ALIASES[raw] ?? CALL_TYPE_ALIASES[lower];
  if (aliased) return aliased;
  if (VALID_STANDARD_CALL_TYPES.has(raw)) return raw;
  if (model && model !== 'unknown') return 'LLM call';
  return raw;
}

const VALID_STANDARD_CALL_TYPES = new Set(['LLM call', 'Tool call', 'Vision call', 'MCP call']);

export interface EdgeTelemetryInput {
  tokens_in?: number;
  tokens_out?: number;
  cost_usd?: number;
  latency_ms?: number;
  model?: string;
  call_type?: string;
  input_preview?: string;
  output_preview?: string;
}

export function enrichEdgeTelemetry(
  edge: EdgeTelemetryInput,
  cfg: Required<TelemetryConfig>,
  providerUsage?: ProviderUsage | null,
): EdgeTelemetryInput {
  const out: EdgeTelemetryInput = { ...edge };
  out.call_type = normalizeCallType(out.call_type, out.model ?? 'unknown');

  let tokensIn = out.tokens_in ?? 0;
  let tokensOut = out.tokens_out ?? 0;
  let latencyMs = out.latency_ms ?? 0;
  let model = out.model ?? 'unknown';

  const missingUsage = tokensIn === 0 && tokensOut === 0 && (out.cost_usd ?? 0) === 0;

  if (missingUsage && cfg.attachProviderUsage && providerUsage) {
    tokensIn = providerUsage.tokens_in;
    tokensOut = providerUsage.tokens_out;
    latencyMs = latencyMs || providerUsage.latency_ms;
    if (model === 'unknown' || !model) model = providerUsage.model;
  }

  if (cfg.enrichMissingUsage) {
    if (tokensIn === 0) {
      tokensIn = estimateTokens(out.input_preview, cfg.charsPerToken);
    }
    if (tokensOut === 0) {
      tokensOut = estimateTokens(out.output_preview, cfg.charsPerToken);
    }
  }

  out.tokens_in = tokensIn;
  out.tokens_out = tokensOut;
  out.latency_ms = latencyMs;
  out.model = model;

  if ((out.cost_usd ?? 0) === 0 && (tokensIn > 0 || tokensOut > 0)) {
    out.cost_usd = estimateCostUsd(model, tokensIn, tokensOut, cfg.modelPricing);
  }

  return out;
}

export function providerUsageFromAnthropic(
  response: {
    model?: string;
    usage?: { input_tokens?: number; output_tokens?: number };
  },
  latencyMs: number,
): ProviderUsage {
  return {
    model: response.model ?? 'unknown',
    tokens_in: response.usage?.input_tokens ?? 0,
    tokens_out: response.usage?.output_tokens ?? 0,
    latency_ms: latencyMs,
    captured_at_ms: Date.now(),
  };
}

export function providerUsageFromOpenAi(
  response: {
    model?: string;
    usage?: { prompt_tokens?: number; completion_tokens?: number };
  },
  latencyMs: number,
): ProviderUsage {
  return {
    model: response.model ?? 'unknown',
    tokens_in: response.usage?.prompt_tokens ?? 0,
    tokens_out: response.usage?.completion_tokens ?? 0,
    latency_ms: latencyMs,
    captured_at_ms: Date.now(),
  };
}

File: packages/sdk-py/src/blamr_sdk/telemetry.py
"""Estimate tokens/cost when agents omit usage on emit_edge."""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Any

DEFAULT_CHARS_PER_TOKEN = 4

DEFAULT_MODEL_PRICING: dict[str, tuple[float, float]] = {
    "claude-sonnet-4-6": (3.0, 15.0),
    "claude-3-5-sonnet-latest": (3.0, 15.0),
    "gpt-4o": (2.5, 10.0),
    "gpt-4o-mini": (0.15, 0.6),
}

CALL_TYPE_ALIASES = {
    "tool_call": "Tool call",
    "llm_call": "LLM call",
    "mcp_call": "MCP call",
    "vision_call": "Vision call",
}


@dataclass
class ProviderUsage:
    model: str
    tokens_in: int
    tokens_out: int
    latency_ms: int


def _env_on(name: str, default: bool = True) -> bool:
    val = os.environ.get(name)
    if val is None:
        return default
    return val not in ("0", "false", "False")


def estimate_tokens(text: str | None, chars_per_token: int = DEFAULT_CHARS_PER_TOKEN) -> int:
    if not text or not str(text).strip():
        return 0
    return max(1, (len(str(text).strip()) + chars_per_token - 1) // chars_per_token)


def _resolve_pricing(model: str, overrides: dict[str, tuple[float, float]] | None = None) -> tuple[float, float] | None:
    pricing = {**DEFAULT_MODEL_PRICING, **(overrides or {})}
    if model in pricing:
        return pricing[model]
    lower = model.lower()
    for key, rates in pricing.items():
        if key.lower() in lower or lower in key.lower():
            return rates
    if "sonnet" in lower:
        return (3.0, 15.0)
    if "haiku" in lower:
        return (0.8, 4.0)
    if "opus" in lower:
        return (15.0, 75.0)
    return None


def estimate_cost_usd(model: str, tokens_in: int, tokens_out: int, overrides: dict[str, tuple[float, float]] | None = None) -> float:
    rates = _resolve_pricing(model, overrides)
    if not rates:
        return 0.0
    inp, out = rates
    return round((tokens_in * inp + tokens_out * out) / 1_000_000, 6)


def normalize_call_type(call_type: str | None, model: str) -> str:
    raw = (call_type or "LLM call").strip()
    lower = raw.lower()
    if lower in ("tool_call", "llm_call") and model and model != "unknown":
        return "LLM call"
    aliased = CALL_TYPE_ALIASES.get(raw) or CALL_TYPE_ALIASES.get(lower)
    if aliased:
        return aliased
    if raw in ("LLM call", "Tool call", "Vision call", "MCP call"):
        return raw
    return "LLM call" if model and model != "unknown" else raw


def enrich_edge_fields(
    fields: dict[str, Any],
    provider_usage: ProviderUsage | None = None,
    *,
    enrich_missing: bool | None = None,
    attach_provider: bool | None = None,
    model_pricing: dict[str, tuple[float, float]] | None = None,
) -> dict[str, Any]:
    enrich = enrich_missing if enrich_missing is not None else _env_on("BLAMR_ENRICH_USAGE")
    attach = attach_provider if attach_provider is not None else _env_on("BLAMR_ATTACH_PROVIDER_USAGE")

    out = dict(fields)
    model = str(out.get("model") or "unknown")
    out["call_type"] = normalize_call_type(out.get("call_type"), model)

    tokens_in = int(out.get("tokens_in") or 0)
    tokens_out = int(out.get("tokens_out") or 0)
    latency_ms = int(out.get("latency_ms") or 0)
    cost_usd = float(out.get("cost_usd") or 0)

    missing = tokens_in == 0 and tokens_out == 0 and cost_usd == 0
    if missing and attach and provider_usage:
        tokens_in = provider_usage.tokens_in
        tokens_out = provider_usage.tokens_out
        latency_ms = latency_ms or provider_usage.latency_ms
        if model in ("unknown", ""):
            model = provider_usage.model

    if enrich:
        if tokens_in == 0:
            tokens_in = estimate_tokens(out.get("input_preview"))
        if tokens_out == 0:
            tokens_out = estimate_tokens(out.get("output_preview"))

    out["tokens_in"] = tokens_in
    out["tokens_out"] = tokens_out
    out["latency_ms"] = latency_ms
    out["model"] = model

    if cost_usd == 0 and (tokens_in > 0 or tokens_out > 0):
        out["cost_usd"] = estimate_cost_usd(model, tokens_in, tokens_out, model_pricing)

    return out
